File: Array/Medium/08_next_permutation.java
package Array.Medium;

/*
 * 31. Next Permutation
 */

class Solution {
    public void nextPermutation(int[] nums) {
        int n = nums.length;
        int endIndex = n-1;
        while (endIndex >= 0) {
            if (endIndex != 0 && nums[endIndex] > nums[endIndex-1]) {
                endIndex--;
                break;
            }
            endIndex--;
        }

        // edge case where the given permutation is last permutation possible
        if (endIndex == -1) {
            endIndex++;
            int end = n-1;
            while (endIndex < end) {
                int t = nums[endIndex];
                nums[endIndex] = nums[end];
                nums[end] = t;
                endIndex++;
                end--;
            }
            return;
        }
        
        // getting the index of element with which we have to interchange nums[endIndex]
        int toExchangeIndex = n-1;
        while (nums[endIndex] >= nums[toExchangeIndex]) {
            toExchangeIndex--;
        }

        int temp = nums[endIndex];
        nums[endIndex] = nums[toExchangeIndex];
        nums[toExchangeIndex] = temp; 

        // reverse the array after endIndex
        endIndex++;
        int newEndIndex = n-1;
        while (endIndex < newEndIndex) {
            // System.out.print(endIndex+" "+ newEndIndex);
            int newTemp = nums[endIndex];
            nums[endIndex] = nums[newEndIndex];
            nums[newEndIndex] = newTemp;
            endIndex++;
            newEndIndex--;
        }

    }
}

File: Array/Medium/08_next_permutation.py
class Solution:
    def nextPermutation(self, N, arr):
        bp = -1  # this breakpoint will point to the index where arr[bp+1:N] will be sorted in dec order
        for i in range(N - 2, -1, -1):
            if arr[i + 1] > arr[i]:
                bp = i
                break

        # if there is no point breakpoint, then that means array is sorted in reverse order, so its next perm will be rev(arr)
        if bp == -1:
            return arr[::-1]

        # now to the right of index bp we need to just greater element than arr[bp] and swap it
        last = N - 1
        while last > bp:
            if arr[last] > arr[bp]:
                arr[last], arr[bp] = arr[bp], arr[last]
                break
            last -= 1

        # reverse the arr[bp+1:], we need to reverse the remaining array because it is sorted in decreasing order, so on reversing will change its sorting
        bp = bp + 1
        last = N - 1
        while bp < last:
            arr[bp], arr[last] = arr[last], arr[bp]
            bp += 1
            last -= 1

        return arr
